Source/WebCore/dom/CustomElementDefaultARIA.cpp

```diff
@@ -40,17 +41,101 @@ void CustomElementDefaultARIA::setValueForAttribute(const QualifiedName& name, c
+            RefPtr element = weakElement.get();
+            if (element && (!element->isConnected() || thisElement.isDescendantOrShadowDescendantOf(element->rootNode()))) {
+                if (idList.length())
+                    idList.append(' ');
+                idList.append(element->attributeWithoutSynchronization(HTMLNames::idAttr));
+            }
+        }
+    }), it->value);
+
+    return *result;
 }
 
 bool CustomElementDefaultARIA::hasAttribute(const QualifiedName& name) const
 {
     return m_map.find(name) != m_map.end();
 }
 
+Element* CustomElementDefaultARIA::elementForAttribute(const Element& thisElement, const QualifiedName& name) const
+{
+    auto it = m_map.find(name);
+    if (it == m_map.end())
+        return nullptr;
+
+    Element* result = nullptr;
+    std::visit(WTF::makeVisitor([&](const AtomString& stringValue) {
+        if (thisElement.isInTreeScope())
+            result = thisElement.treeScope().getElementById(stringValue);
+    }, [&](const WeakPtr<Element, WeakPtrImplWithEventTargetData>& weakElementValue) {
+        RefPtr elementValue = weakElementValue.get();
+        if (elementValue && (!elementValue->isConnected() || thisElement.isDescendantOrShadowDescendantOf(elementValue->rootNode())))
+            result = elementValue.get();
+    }, [&](const Vector<WeakPtr<Element, WeakPtrImplWithEventTargetData>>&) {
+        RELEASE_ASSERT_NOT_REACHED();
+    }), it->value);
+    return result;
+}
+
+void CustomElementDefaultARIA::setElementForAttribute(const QualifiedName& name, Element* element)
+{
+    m_map.set(name, WeakPtr<Element, WeakPtrImplWithEventTargetData> { element });
+}
+
+Vector<RefPtr<Element>> CustomElementDefaultARIA::elementsForAttribute(const Element& thisElement, const QualifiedName& name) const
+{
+    Vector<RefPtr<Element>> result;
+    auto it = m_map.find(name);
+    if (it == m_map.end())
+        return result;
+    std::visit(WTF::makeVisitor([&](const AtomString& stringValue) {
+        SpaceSplitString idList { stringValue, SpaceSplitString::ShouldFoldCase::No };
+        result.reserveCapacity(idList.size());
+        if (thisElement.isInTreeScope()) {
+            for (unsigned i = 0; i < idList.size(); ++i)
+                result.append(thisElement.treeScope().getElementById(idList[i]));
+        }
+    }, [&](const WeakPtr<Element, WeakPtrImplWithEventTargetData>& weakElementValue) {
+        RefPtr element = weakElementValue.get();
+        if (element && (!element->isConnected() || thisElement.isDescendantOrShadowDescendantOf(element->rootNode())))
+            result.append(WTFMove(element));
+    }, [&](const Vector<WeakPtr<Element, WeakPtrImplWithEventTargetData>>& elements) {
+        result.reserveCapacity(elements.size());
+        for (auto& weakElement : elements) {
+            if (RefPtr element = weakElement.get(); element && (!element->isConnected() || thisElement.isDescendantOrShadowDescendantOf(element->rootNode())))
```

Comment:
It would be nice to reuse these checks somehow and avoid having to repeat them in the different methods.

Reply:
Extracted it as a helper function: `isElementVisible`

Reply:
I guess the new changes haven't been pushed to the PR yet, or at least I cannot see them.